File: fraiseql-kotlin/src/test/kotlin/com/fraiseql/security/RoleBasedAccessControlTest.kt
package com.fraiseql.security

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class RoleBasedAccessControlTest {

    @Test
    fun `should create single role requirement`() {
        val config = RoleRequiredBuilder()
            .roles("admin")
            .description("Admin role required")
            .build()

        assertEquals(1, config.roles.size)
        assertTrue(config.roles.contains("admin"))
    }

    @Test
    fun `should create multiple role requirements`() {
        val config = RoleRequiredBuilder()
            .roles("manager", "director")
            .description("Manager or director required")
            .build()

        assertEquals(2, config.roles.size)
        assertTrue(config.roles.contains("manager"))
        assertTrue(config.roles.contains("director"))
    }

    @Test
    fun `should create roles from array`() {
        val config = RoleRequiredBuilder()
            .rolesArray(listOf("viewer", "editor", "admin"))
            .description("Multiple roles via array")
            .build()

        assertEquals(3, config.roles.size)
    }

    @Test
    fun `should support ANY matching strategy`() {
        val config = RoleRequiredBuilder()
            .roles("manager", "director")
            .strategy(RoleMatchStrategy.Any)
            .description("User needs at least one role")
            .build()

        assertEquals(RoleMatchStrategy.Any, config.strategy)
    }

    @Test
    fun `should support ALL matching strategy`() {
        val config = RoleRequiredBuilder()
            .roles("admin", "auditor")
            .strategy(RoleMatchStrategy.All)
            .description("User needs all roles")
            .build()

        assertEquals(RoleMatchStrategy.All, config.strategy)
    }

    @Test
    fun `should support EXACTLY matching strategy`() {
        val config = RoleRequiredBuilder()
            .roles("admin")
            .strategy(RoleMatchStrategy.Exactly)
            .description("User must have exactly these roles")
            .build()

        assertEquals(RoleMatchStrategy.Exactly, config.strategy)
    }

    @Test
    fun `should support role hierarchy`() {
        val config = RoleRequiredBuilder()
            .roles("user")
            .hierarchy(true)
            .description("Role hierarchy enabled")
            .build()

        assertTrue(config.hierarchy)
    }

    @Test
    fun `should support role inheritance`() {
        val config = RoleRequiredBuilder()
            .roles("editor")
            .inherit(true)
            .description("Inherit role requirements")
            .build()

        assertTrue(config.inherit)
    }

    @Test
    fun `should support operation-specific rules`() {
        val config = RoleRequiredBuilder()
            .roles("admin")
            .operations("delete,create")
            .description("Admin for destructive operations")
            .build()

        assertEquals("delete,create", config.operations)
    }

    @Test
    fun `should support caching`() {
        val config = RoleRequiredBuilder()
            .roles("viewer")
            .cacheable(true)
            .cacheDurationSeconds(1800)
            .build()

        assertTrue(config.cacheable)
        assertEquals(1800, config.cacheDurationSeconds)
    }

    @Test
    fun `should support custom error message`() {
        val config = RoleRequiredBuilder()
            .roles("admin")
            .errorMessage("You must be an administrator to access this resource")
            .build()

        assertEquals("You must be an administrator to access this resource", config.errorMessage)
    }

    @Test
    fun `should support fluent chaining`() {
        val config = RoleRequiredBuilder()
            .roles("manager", "director")
            .strategy(RoleMatchStrategy.Any)
            .hierarchy(true)
            .description("Manager or director with hierarchy")
            .errorMessage("Insufficient role")
            .operations("read,update")
            .inherit(false)
            .cacheable(true)
            .cacheDurationSeconds(900)
            .build()

        assertEquals(2, config.roles.size)
        assertEquals(RoleMatchStrategy.Any, config.strategy)
        assertTrue(config.hierarchy)
        assertFalse(config.inherit)
        assertEquals(900, config.cacheDurationSeconds)
    }

    @Test
    fun `should support admin pattern`() {
        val config = RoleRequiredBuilder()
            .roles("admin")
            .strategy(RoleMatchStrategy.Exactly)
            .hierarchy(true)
            .description("Full admin access with hierarchy")
            .build()

        assertEquals(1, config.roles.size)
        assertTrue(config.hierarchy)
    }

    @Test
    fun `should support manager pattern`() {
        val config = RoleRequiredBuilder()
            .roles("manager", "director", "executive")
            .strategy(RoleMatchStrategy.Any)
            .description("Management tier access")
            .operations("read,create,update")
            .build()

        assertEquals(3, config.roles.size)
        assertEquals("read,create,update", config.operations)
    }

    @Test
    fun `should support data scientist pattern`() {
        val config = RoleRequiredBuilder()
            .roles("data_scientist", "analyst")
            .strategy(RoleMatchStrategy.Any)
            .description("Data access for scientists and analysts")
            .operations("read")
            .build()

        assertEquals(2, config.roles.size)
    }

    @Test
    fun `should support annotation syntax`() {
        @RoleRequired(
            roles = ["admin"],
            description = "Admin access required"
        )
        class AdminPanel

        val annotation = AdminPanel::class.java.getAnnotation(RoleRequired::class.java)
        assertEquals(1, annotation?.roles?.size)
    }

    @Test
    fun `should support annotation with strategy`() {
        @RoleRequired(
            roles = ["manager", "director"],
            strategy = "any",
            description = "Management access"
        )
        class SalaryData

        val annotation = SalaryData::class.java.getAnnotation(RoleRequired::class.java)
        assertEquals("any", annotation?.strategy)
    }

    @Test
    fun `should support annotation with all parameters`() {
        @RoleRequired(
            roles = ["admin", "auditor"],
            strategy = "all",
            hierarchy = true,
            description = "Full admin with auditor",
            errorMessage = "Insufficient privileges",
            operations = "delete,create",
            inherit = false,
            cacheable = true,
            cacheDurationSeconds = 1200
        )
        class ComplexRoleRequirement

        val annotation = ComplexRoleRequirement::class.java.getAnnotation(RoleRequired::class.java)
        assertTrue(annotation?.hierarchy ?: false)
    }

    @Test
    fun `should create multiple roles with different strategies`() {
        val any = RoleRequiredBuilder()
            .roles("editor", "contributor")
            .strategy(RoleMatchStrategy.Any)
            .build()

        val all = RoleRequiredBuilder()
            .roles("editor", "reviewer")
            .strategy(RoleMatchStrategy.All)
            .build()

        val exactly = RoleRequiredBuilder()
            .roles("admin")
            .strategy(RoleMatchStrategy.Exactly)
            .build()

        assertEquals(RoleMatchStrategy.Any, any.strategy)
        assertEquals(RoleMatchStrategy.All, all.strategy)
        assertEquals(RoleMatchStrategy.Exactly, exactly.strategy)
    }
}

File: fraiseql-nodejs/tests/role-based-access-control.test.ts
import { describe, it, expect } from '@jest/globals';
import {
  RoleRequiredBuilder,
  RoleMatchStrategy,
  RoleRequired,
} from '../src/security';

describe('Role-Based Access Control', () => {
  it('should create single role requirement', () => {
    const config = new RoleRequiredBuilder()
      .roles('admin')
      .description('Admin role required')
      .build();

    expect(config.roles).toHaveLength(1);
    expect(config.roles).toContain('admin');
  });

  it('should create multiple role requirements', () => {
    const config = new RoleRequiredBuilder()
      .roles('manager', 'director')
      .description('Manager or director required')
      .build();

    expect(config.roles).toHaveLength(2);
    expect(config.roles).toContain('manager');
    expect(config.roles).toContain('director');
  });

  it('should create roles from array', () => {
    const config = new RoleRequiredBuilder()
      .rolesArray(['viewer', 'editor', 'admin'])
      .description('Multiple roles via array')
      .build();

    expect(config.roles).toHaveLength(3);
  });

  it('should support ANY matching strategy', () => {
    const config = new RoleRequiredBuilder()
      .roles('manager', 'director')
      .strategy(RoleMatchStrategy.ANY)
      .description('User needs at least one role')
      .build();

    expect(config.strategy).toBe(RoleMatchStrategy.ANY);
  });

  it('should support ALL matching strategy', () => {
    const config = new RoleRequiredBuilder()
      .roles('admin', 'auditor')
      .strategy(RoleMatchStrategy.ALL)
      .description('User needs all roles')
      .build();

    expect(config.strategy).toBe(RoleMatchStrategy.ALL);
  });

  it('should support EXACTLY matching strategy', () => {
    const config = new RoleRequiredBuilder()
      .roles('admin')
      .strategy(RoleMatchStrategy.EXACTLY)
      .description('User must have exactly these roles')
      .build();

    expect(config.strategy).toBe(RoleMatchStrategy.EXACTLY);
  });

  it('should support role hierarchy', () => {
    const config = new RoleRequiredBuilder()
      .roles('user')
      .hierarchy(true)
      .description('Role hierarchy enabled')
      .build();

    expect(config.hierarchy).toBe(true);
  });

  it('should support role inheritance', () => {
    const config = new RoleRequiredBuilder()
      .roles('editor')
      .inherit(true)
      .description('Inherit role requirements')
      .build();

    expect(config.inherit).toBe(true);
  });

  it('should support operation-specific rules', () => {
    const config = new RoleRequiredBuilder()
      .roles('admin')
      .operations('delete,create')
      .description('Admin for destructive operations')
      .build();

    expect(config.operations).toBe('delete,create');
  });

  it('should support caching', () => {
    const config = new RoleRequiredBuilder()
      .roles('viewer')
      .cacheable(true)
      .cacheDurationSeconds(1800)
      .build();

    expect(config.cacheable).toBe(true);
    expect(config.cacheDurationSeconds).toBe(1800);
  });

  it('should support custom error message', () => {
    const config = new RoleRequiredBuilder()
      .roles('admin')
      .errorMessage('You must be an administrator to access this resource')
      .build();

    expect(config.errorMessage).toBe(
      'You must be an administrator to access this resource'
    );
  });

  it('should support fluent chaining', () => {
    const config = new RoleRequiredBuilder()
      .roles('manager', 'director')
      .strategy(RoleMatchStrategy.ANY)
      .hierarchy(true)
      .description('Manager or director with hierarchy')
      .errorMessage('Insufficient role')
      .operations('read,update')
      .inherit(false)
      .cacheable(true)
      .cacheDurationSeconds(900)
      .build();

    expect(config.roles).toHaveLength(2);
    expect(config.strategy).toBe(RoleMatchStrategy.ANY);
    expect(config.hierarchy).toBe(true);
    expect(config.inherit).toBe(false);
    expect(config.cacheDurationSeconds).toBe(900);
  });

  it('should support admin pattern', () => {
    const config = new RoleRequiredBuilder()
      .roles('admin')
      .strategy(RoleMatchStrategy.EXACTLY)
      .hierarchy(true)
      .description('Full admin access with hierarchy')
      .build();

    expect(config.roles).toHaveLength(1);
    expect(config.hierarchy).toBe(true);
  });

  it('should support manager pattern', () => {
    const config = new RoleRequiredBuilder()
      .roles('manager', 'director', 'executive')
      .strategy(RoleMatchStrategy.ANY)
      .description('Management tier access')
      .operations('read,create,update')
      .build();

    expect(config.roles).toHaveLength(3);
    expect(config.operations).toBe('read,create,update');
  });

  it('should support data scientist pattern', () => {
    const config = new RoleRequiredBuilder()
      .roles('data_scientist', 'analyst')
      .strategy(RoleMatchStrategy.ANY)
      .description('Data access for scientists and analysts')
      .operations('read')
      .build();

    expect(config.roles).toHaveLength(2);
  });

  it('should support decorator syntax', () => {
    @RoleRequired({
      roles: ['admin'],
      description: 'Admin access required',
    })
    class AdminPanel {
      content: string;
    }

    expect(AdminPanel).toBeDefined();
  });

  it('should support decorator with strategy', () => {
    @RoleRequired({
      roles: ['manager', 'director'],
      strategy: RoleMatchStrategy.ANY,
      description: 'Management access',
    })
    class SalaryData {
      employeeId: string;
      salary: number;
    }

    expect(SalaryData).toBeDefined();
  });

  it('should support decorator with all parameters', () => {
    @RoleRequired({
      roles: ['admin', 'auditor'],
      strategy: RoleMatchStrategy.ALL,
      hierarchy: true,
      description: 'Full admin with auditor',
      errorMessage: 'Insufficient privileges',
      operations: 'delete,create',
      inherit: false,
      cacheable: true,
      cacheDurationSeconds: 1200,
    })
    class ComplexRoleRequirement {
      data: string;
    }

    expect(ComplexRoleRequirement).toBeDefined();
  });

  it('should create multiple roles with different strategies', () => {
    const any = new RoleRequiredBuilder()
      .roles('editor', 'contributor')
      .strategy(RoleMatchStrategy.ANY)
      .build();

    const all = new RoleRequiredBuilder()
      .roles('editor', 'reviewer')
      .strategy(RoleMatchStrategy.ALL)
      .build();

    const exactly = new RoleRequiredBuilder()
      .roles('admin')
      .strategy(RoleMatchStrategy.EXACTLY)
      .build();

    expect(any.strategy).toBe(RoleMatchStrategy.ANY);
    expect(all.strategy).toBe(RoleMatchStrategy.ALL);
    expect(exactly.strategy).toBe(RoleMatchStrategy.EXACTLY);
  });
});
